overlord/aspectstate/aspectstate_test.go

```diff
@@ -196,3 +197,146 @@ func (s *aspectTestSuite) TestNewTransactionNoState(c *C) {
 		c.Assert(value, Equals, "bar")
 	}
 }
+
+type filterSampleSuite struct {
+	state *state.State
+	bag   aspects.JSONDataBag
+}
+
+var _ = Suite(&filterSampleSuite{})
+
+func (s *filterSampleSuite) SetUpTest(c *C) {
+	var raw map[string]json.RawMessage
+	err := json.Unmarshal([]byte(`{
+		"snaps": {
+			"test-snapd-sh": {
+				"name":   "test-snapd-sh",
+				"status": "installed"
+			},
+			"core20": {
+				"name":   "core20",
+				"status": "active"
+			},
+			"snapcraft": {
+				"name":   "snapcraft",
+				"status": "active"
+			},
+			"firefox": {
+				"name":   "firefox",
+				"status": "active"
+			},
+			"snapd": {
+				"name":   "snapd",
+				"status": "active"
+			},
+			"vlc": {
+				"name":   "vlc",
+				"status": "inactive"
+			},
+			"spotify": {
+				"name":   "spotify",
+				"status": "failed"
+			},
+			"discord": {
+				"name":   "discord",
+				"status": "active"
+			},
+			"shellcheck": {
+				"name":   "shellcheck",
+				"status": "active"
+			},
+			"htop": {
+				"name":   "htop",
+				"status": "inactive"
+			}
+		}
+	}`), &raw)
+	c.Assert(err, IsNil)
+
+	s.state = overlord.Mock().State()
+	s.bag = aspects.JSONDataBag(raw)
+}
+
+func (s *filterSampleSuite) TestQueryNoFilters(c *C) {
+	var res interface{}
+	err := aspectstate.QueryAspect(s.bag, "acc", "bundle", "asp", "snaps", "", &res)
+	c.Assert(err, IsNil)
+	// returns all snaps
+	obj, ok := res.(map[string]json.RawMessage)
+	c.Assert(ok, Equals, true)
+	c.Assert(obj, HasLen, 10)
+
+	discordRaw, ok := obj["discord"]
+	c.Assert(ok, Equals, true)
+
+	var discord map[string]json.RawMessage
+	err = json.Unmarshal(discordRaw, &discord)
+	c.Assert(err, IsNil)
+
+	assertField(c, discord, "name", "discord")
+	assertField(c, discord, "status", "active")
+
+	htopRaw, ok := obj["htop"]
+	c.Assert(ok, Equals, true)
+
+	var htop map[string]json.RawMessage
+	err = json.Unmarshal(htopRaw, &htop)
+	c.Assert(err, IsNil)
+
+	assertField(c, htop, "name", "htop")
+	assertField(c, htop, "status", "inactive")
+}
+
+func (s *filterSampleSuite) TestQueryFilterNameWithParameter(c *C) {
+	var res interface{}
+	err := aspectstate.QueryAspect(s.bag, "acc", "bundle", "asp", "snaps", "name=firefox", &res)
+	c.Assert(err, IsNil)
+	obj, ok := res.(map[string]json.RawMessage)
```

Comment:
shouldn't this produce an output like:

```
{
  "firefox": {
    ...
  }
}
```

?

Reply:
This is the main point where I have doubts. For the output to be that, in the path/storage `snaps.{name}[.status={status}]`, the `{name}` part can be a sub-key (if used with a request like `snaps.firefox` or a filter, if the placeholder is met by a query parameter instead of a subkey in the request. That seemed strange because the [.status={status}] is also filtering by field so. 
Assuming that behaviour, if we have `request: "snaps", query: "name=firefox,status=active"` (if we wanted to get a snap but only if it's active, for instance), then the query seems a bit inconsistent. The `snaps` part gives us `{"firefox": {...}, "snapd": {...}, ...}` and "name=firefox" would filter that object to get only `{"firefox": {...}}`, then we can't really apply the field filter because the only one is "firefox" which maps to the snap. It also seems a bit strange because, unlike the case of the field filter, there's only one value that can meet this filter. The only possible values are an object with the firefox snap or an empty object. 
In the current implementation, `request: "snaps.firefox", query: ""` and `request: "snaps", query: "name=firefox"` both return `{"name": "firefox", "status": "active"}` (they both act fill the placeholder and use it as a key). In both cases, we can still have a query that includes a field filter (like `"status=active"`) because the fields are at the top level.

Reply:
I'm a bit confused by the last comment,  order of filters shouldn't matter are if it matters internally we should make it so that things commute.

> The only possible values are an object with the firefox snap or an empty object.

but that seems reasonable to expect from "name=firefox,status=active"

as I said in the standup maybe we want for now all results to look like:

{ "snaps":  {"foo": ...}}

Reply:
I've updated this now to return the whole object mapped to by the key